CritBit/src/org/zoodb/index/critbit/CritBit64COW.java

```diff
@@ -422,82 +441,89 @@ public V remove(long key) {
+            Node<V> parent = null;
+            boolean isParentHigh = false;
+            long prefix = info.rootKey;
+            while (doesPrefixMatch(n.posDiff, key, prefix)) {
+                //perform copy on write
+                n = new Node<V>(n);
+                if (parent != null) {
+                    if (!isParentHigh) {
+                        parent.lo = n;
+                    } else {
+                        parent.hi = n;
                     }
-                    //match! --> delete node
-                    //replace data in parent node
-                    updateParentAfterRemove(parent, n.loPost, n.loVal, n.lo, isParentHigh);
-                    return n.hiVal;
+                } else {
+                    newInfo.root = n;
                 }
-            } else {
-                if (n.lo != null) {
-                    isParentHigh = false;
-                    prefix = n.loPost;
-                    parent = n;
-                    n = n.lo;
-                    continue;
+
+                //prefix matches, so now we check sub-nodes and postfixes
+                if (BitTools.getBit(key, n.posDiff)) {
+                    if (n.hi != null) {
+                        isParentHigh = true;
+                        prefix = n.hiPost;
+                        parent = n;
+                        n = n.hi;
+                        continue;
+                    } else {
+                        if (key != n.hiPost) {
+                            return null;
+                        }
+                        //match! --> delete node
+                        //replace data in parent node
+                        updateParentAfterRemove(newInfo, parent, n.loPost, n.loVal, n.lo, isParentHigh);
+                        return n.hiVal;
+                    }
                 } else {
-                    if (key != n.loPost) {
-                        return null;
+                    if (n.lo != null) {
+                        isParentHigh = false;
+                        prefix = n.loPost;
+                        parent = n;
+                        n = n.lo;
+                        continue;
+                    } else {
+                        if (key != n.loPost) {
+                            return null;
+                        }
+                        //match! --> delete node
+                        //replace data in parent node
+                        //for new prefixes...
+                        updateParentAfterRemove(newInfo, parent, n.hiPost, n.hiVal, n.hi, isParentHigh);
+                        return n.loVal;
                     }
-                    //match! --> delete node
-                    //replace data in parent node
-                    //for new prefixes...
-                    updateParentAfterRemove(parent, n.hiPost, n.hiVal, n.hi, isParentHigh);
-                    return n.loVal;
                 }
             }
+        } finally {
+            this.info = newInfo;
```

Comment:
Well, if there should be no assignment in case of an exception, I can just remove the try - finally block and perform the this.info = newInfo assignment on the last line of the doRemove()/doPut() methods.

Your solution is also possible, as it doesn't require many modifications either. Let me know which of these sounds better to you and I will peform the changes.